usr/lib/ica_s390_stdll/ica_specific.c

```diff
@@ -3124,6 +3125,364 @@ static CK_RV os_specific_rsa_decrypt(STDLL_TokData_t *tokdata,
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    (*out_data_len) = out_len;
+
+    /* AAD only processed in first update sequence,
+     * mark it empty for all subsequent calls
+     */
+    context->ulAlen = 0;
+
+done:
+    if (buffer)
+        free(buffer);
+
+    return rc;
+}
+
+CK_RV new_gcm_specific_aes_gcm_final(STDLL_TokData_t *tokdata, SESSION *sess,
+                                     ENCR_DECR_CONTEXT *ctx, CK_BYTE *out_data,
+                                     CK_ULONG *out_data_len, CK_BYTE encrypt)
+{
+    CK_RV rc = CKR_OK;
+    AES_GCM_CONTEXT *context = NULL;
+    CK_GCM_PARAMS *aes_gcm_param = NULL;
+    kma_ctx *gcm_ctx = NULL;
+    CK_BYTE *final_tag_data, *buffer = NULL;;
+    CK_ULONG tag_data_len;
+    CK_BYTE tmp_tag[16];
+
+    UNUSED(tokdata);
+    UNUSED(sess);
+    UNUSED(encrypt);
+
+    context = (AES_GCM_CONTEXT *)ctx->context;
+    gcm_ctx = (kma_ctx *)context->ulClen;
+    aes_gcm_param = (CK_GCM_PARAMS *)ctx->mech.pParameter;
+    tag_data_len = (aes_gcm_param->ulTagBits + 7) / 8;
+
+    if (encrypt) {
+        if (context->len != 0) {
+            buffer = (CK_BYTE *)malloc(AES_BLOCK_SIZE);
+            if (!buffer) {
+                TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+                rc = CKR_HOST_MEMORY;
+                goto done;
+            }
+            memcpy(buffer, context->data, context->len);
+
+            /* Perform final update with rest of data to calculate final tag */
+            rc = ica_aes_gcm_kma_update(buffer, out_data, context->len,
+                                        NULL, 0, 1,1, gcm_ctx);
+            if (rc != 0) {
+                TRACE_ERROR("ica_aes_gcm_kma_update failed, rc=%ld\n",rc);
+                rc = CKR_FUNCTION_FAILED;
+                goto done;
+            }
+
+            *out_data_len = context->len + tag_data_len;
+        } else {
+            /* Perform final update without data to calculate final tag */
+            rc = ica_aes_gcm_kma_update(NULL, NULL, 0, NULL, 0, 1, 1, gcm_ctx);
+            *out_data_len = tag_data_len;
+        }
+
+        TRACE_DEVEL("GCM Final: context->len=%ld, tag_data_len=%ld, "
+                    "out_data_len=%ld\n",
+                    context->len, tag_data_len, *out_data_len);
+
+        rc = ica_aes_gcm_kma_get_tag(tmp_tag, tag_data_len, gcm_ctx);
+        if (rc != 0) {
+            TRACE_ERROR("ica_aes_gcm_kma_get_tag failed, rc=%ld.\n", rc);
+            rc = CKR_FUNCTION_FAILED;
+            goto done;
+        }
+
+        memcpy(out_data + context->len, tmp_tag, tag_data_len);
+
+    } else {
+        /* decrypt */
+        if (context->len > tag_data_len) {
+            buffer = (CK_BYTE *)malloc(AES_BLOCK_SIZE);
+            if (!buffer) {
+                TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+                rc = CKR_HOST_MEMORY;
+                goto done;
+            }
+            memcpy(buffer, context->data, context->len - tag_data_len);
```

Comment:
Again, why copying into a buffer? you can use context->data right away with ica_aes_gcm_kma_update below.

Reply:
ok, done.